File: packages/coding-agent/src/core/bash-executor.ts
/**
 * Bash command execution with streaming support and cancellation.
 *
 * This module provides a unified bash execution implementation used by:
 * - AgentSession.executeBash() for interactive and RPC modes
 * - Direct calls from modes that need bash execution
 */

import stripAnsi from "strip-ansi";
import { sanitizeBinaryOutput } from "../utils/shell.js";
import { createTempOutputCapture } from "./temp-output-capture.js";
import { type BashOperations, createLocalBashOperations } from "./tools/bash.js";
import { DEFAULT_MAX_BYTES, truncateTail } from "./tools/truncate.js";

// ============================================================================
// Types
// ============================================================================

export interface BashExecutorOptions {
	/** Callback for streaming output chunks (already sanitized) */
	onChunk?: (chunk: string) => void;
	/** AbortSignal for cancellation */
	signal?: AbortSignal;
}

export interface BashResult {
	/** Combined stdout + stderr output (sanitized, possibly truncated) */
	output: string;
	/** Process exit code (undefined if killed/cancelled) */
	exitCode: number | undefined;
	/** Whether the command was cancelled via signal */
	cancelled: boolean;
	/** Whether the output was truncated */
	truncated: boolean;
	/** Path to temp file containing full output (if output exceeded truncation threshold) */
	fullOutputPath?: string;
}

// ============================================================================
// Implementation
// ============================================================================

/**
 * Execute a bash command with optional streaming and cancellation support.
 *
 * Uses the same local BashOperations backend as createBashTool() so interactive
 * user bash and tool-invoked bash share the same process spawning behavior.
 * Sanitization, newline normalization, temp-file capture, and truncation still
 * happen in executeBashWithOperations(), so reusing the local backend does not
 * change output processing behavior.
 *
 * @param command - The bash command to execute
 * @param options - Optional streaming callback and abort signal
 * @returns Promise resolving to execution result
 */
export function executeBash(command: string, options?: BashExecutorOptions): Promise<BashResult> {
	return executeBashWithOperations(command, process.cwd(), createLocalBashOperations(), options);
}

/**
 * Execute a bash command using custom BashOperations.
 * Used for remote execution (SSH, containers, etc.).
 */
export async function executeBashWithOperations(
	command: string,
	cwd: string,
	operations: BashOperations,
	options?: BashExecutorOptions,
): Promise<BashResult> {
	const outputChunks: string[] = [];
	let outputBytes = 0;
	const maxOutputBytes = DEFAULT_MAX_BYTES * 2;

	let tempFilePath: string | undefined;
	let tempOutputCapture: ReturnType<typeof createTempOutputCapture> | undefined;
	let totalBytes = 0;

	const decoder = new TextDecoder();

	const onData = (data: Buffer) => {
		totalBytes += data.length;

		// Sanitize: strip ANSI, replace binary garbage, normalize newlines
		const text = sanitizeBinaryOutput(stripAnsi(decoder.decode(data, { stream: true }))).replace(/\r/g, "");

		// Start writing to temp file if exceeds threshold
		if (totalBytes > DEFAULT_MAX_BYTES && !tempOutputCapture) {
			tempOutputCapture = createTempOutputCapture("pi-bash-", ".log");
			tempFilePath = tempOutputCapture?.path;
			for (const chunk of outputChunks) {
				tempOutputCapture?.write(chunk);
			}
		}

		tempOutputCapture?.write(text);

		// Keep rolling buffer
		outputChunks.push(text);
		outputBytes += text.length;
		while (outputBytes > maxOutputBytes && outputChunks.length > 1) {
			const removed = outputChunks.shift()!;
			outputBytes -= removed.length;
		}

		// Stream to callback
		if (options?.onChunk) {
			options.onChunk(text);
		}
	};

	try {
		const result = await operations.exec(command, cwd, {
			onData,
			signal: options?.signal,
		});

		tempOutputCapture?.close();

		const fullOutput = outputChunks.join("");
		const truncationResult = truncateTail(fullOutput);
		const cancelled = options?.signal?.aborted ?? false;

		return {
			output: truncationResult.truncated ? truncationResult.content : fullOutput,
			exitCode: cancelled ? undefined : (result.exitCode ?? undefined),
			cancelled,
			truncated: truncationResult.truncated,
			fullOutputPath: tempFilePath,
		};
	} catch (err) {
		tempOutputCapture?.close();

		// Check if it was an abort
		if (options?.signal?.aborted) {
			const fullOutput = outputChunks.join("");
			const truncationResult = truncateTail(fullOutput);
			return {
				output: truncationResult.truncated ? truncationResult.content : fullOutput,
				exitCode: undefined,
				cancelled: true,
				truncated: truncationResult.truncated,
				fullOutputPath: tempFilePath,
			};
		}

		throw err;
	}
}

File: python/src/pi_mono/coding_agent/core/bash_executor.py
"""Bash command execution with streaming support and cancellation."""

from __future__ import annotations

import os
import secrets
import tempfile
from dataclasses import dataclass
from typing import Any, Callable, Protocol

from pi_mono.coding_agent.core.tools.truncate import DEFAULT_MAX_BYTES, truncateTail
from pi_mono.utils.ansi import strip_ansi
from pi_mono.utils.shell import sanitize_binary_output


class BashOperations(Protocol):
    async def exec(
        self,
        command: str,
        cwd: str,
        *,
        on_data: Callable[[bytes], None],
        signal: Any = None,
        timeout: float | None = None,
        env: dict[str, str] | None = None,
    ) -> dict[str, int | None]: ...


@dataclass
class BashExecutorOptions:
    on_chunk: Callable[[str], None] | None = None
    signal: Any = None


@dataclass
class BashResult:
    output: str
    exit_code: int | None
    cancelled: bool
    truncated: bool
    full_output_path: str | None = None


async def execute_bash_with_operations(
    command: str,
    cwd: str,
    operations: BashOperations,
    options: BashExecutorOptions | None = None,
) -> BashResult:
    output_chunks: list[str] = []
    output_bytes = 0
    max_output_bytes = DEFAULT_MAX_BYTES * 2
    temp_file_path: str | None = None
    temp_file_handle: Any = None
    total_bytes = 0

    def ensure_temp_file() -> None:
        nonlocal temp_file_path, temp_file_handle
        if temp_file_path is not None:
            return
        temp_file_path = os.path.join(tempfile.gettempdir(), f"pi-bash-{secrets.token_hex(8)}.log")
        temp_file_handle = open(temp_file_path, "w", encoding="utf-8")
        for chunk in output_chunks:
            temp_file_handle.write(chunk)

    def on_data(data: bytes) -> None:
        nonlocal output_bytes, total_bytes
        total_bytes += len(data)
        text = sanitize_binary_output(strip_ansi(data.decode("utf-8", errors="replace"))).replace(
            "\r", ""
        )

        if total_bytes > DEFAULT_MAX_BYTES:
            ensure_temp_file()

        if temp_file_handle is not None:
            temp_file_handle.write(text)

        output_chunks.append(text)
        output_bytes += len(text)
        while output_bytes > max_output_bytes and len(output_chunks) > 1:
            removed = output_chunks.pop(0)
            output_bytes -= len(removed)

        if options and options.on_chunk:
            options.on_chunk(text)

    try:
        result = await operations.exec(
            command, cwd, on_data=on_data, signal=options.signal if options else None
        )
        full_output = "".join(output_chunks)
        truncation_result = truncateTail(full_output)
        if truncation_result["truncated"]:
            ensure_temp_file()
        if temp_file_handle is not None:
            temp_file_handle.close()
        cancelled = bool(
            options and options.signal is not None and getattr(options.signal, "aborted", False)
        )
        return BashResult(
            output=truncation_result["content"] if truncation_result["truncated"] else full_output,
            exit_code=None if cancelled else result.get("exitCode"),
            cancelled=cancelled,
            truncated=bool(truncation_result["truncated"]),
            full_output_path=temp_file_path,
        )
    except Exception as err:
        if options and options.signal is not None and getattr(options.signal, "aborted", False):
            full_output = "".join(output_chunks)
            truncation_result = truncateTail(full_output)
            if truncation_result["truncated"]:
                ensure_temp_file()
            if temp_file_handle is not None:
                temp_file_handle.close()
            return BashResult(
                output=(
                    truncation_result["content"] if truncation_result["truncated"] else full_output
                ),
                exit_code=None,
                cancelled=True,
                truncated=bool(truncation_result["truncated"]),
                full_output_path=temp_file_path,
            )
        if temp_file_handle is not None:
            temp_file_handle.close()
        raise err
